packages/teko/src/Tpetra/Teko_BlockingTpetra.cpp

```diff
@@ -164,148 +165,193 @@ void many2one(Tpetra::MultiVector<ST, LO, GO, NT>& one,
-
-    A->getLocalRowCopy(lid, indices, values, numEntries);
-
-    LO numOwnedCols = 0;
-    auto data       = local2ContigGIDs.getLocalViewHost(Tpetra::Access::ReadOnly);
-    for (size_t localCol = 0; localCol < numEntries; localCol++) {
-      TEUCHOS_ASSERT(indices(localCol) > -1);
-
-      // if global id is not owned by this column
-
-      GO gid = data(indices(localCol), 0);
-      if (gid == -1) continue;  // in contiguous row
-
-      colIndices[numOwnedCols] = gid;
-      colValues[numOwnedCols]  = values(localCol);
-      numOwnedCols++;
-    }
-
-    // insert it into the new matrix
-    mat->insertGlobalValues(contigRow, Teuchos::ArrayView<GO>(colIndices.data(), numOwnedCols),
-                            Teuchos::ArrayView<ST>(colValues.data(), numOwnedCols));
-  }
-
-  // fill it and automagically optimize the storage
-  mat->fillComplete(colMap, rowMap);
+  TEUCHOS_ASSERT(prefixSumHost(numMyRows) == totalNumOwnedCols)
+  Kokkos::deep_copy(prefixSumEntriesPerRow, prefixSumHost);
+
+  using device_type        = typename NT::device_type;
+  auto uniqueColumnIndices = Kokkos::View<GO*, device_type>(
+      Kokkos::ViewAllocateWithoutInitializing("uniqueColumnIndices"), totalNumOwnedCols);
+  auto columnIndices = Kokkos::View<GO*>(Kokkos::ViewAllocateWithoutInitializing("columnIndices"),
+                                         totalNumOwnedCols);
+  auto values =
+      Kokkos::View<ST*>(Kokkos::ViewAllocateWithoutInitializing("values"), totalNumOwnedCols);
+
+  Kokkos::parallel_for(
+      Kokkos::RangePolicy<Kokkos::Schedule<Kokkos::Dynamic>, matrix_execution_space>(0, numMyRows),
+      KOKKOS_LAMBDA(const LO localRow) {
+        GO globalRow             = gRowMap_dev.getGlobalElement(localRow);
+        LO lid                   = A_rowmap_dev.getLocalElement(globalRow);
+        const auto sparseRowView = A_dev.row(lid);
+
+        LO colId      = 0;
+        LO colIdStart = prefixSumEntriesPerRow[localRow];
+        for (auto localCol = 0; localCol < sparseRowView.length; localCol++) {
+          GO gid = data(sparseRowView.colidx(localCol), 0);
+          if (gid == invalid) continue;
+          auto value                              = sparseRowView.value(localCol);
+          uniqueColumnIndices(colId + colIdStart) = gid;
+          values(colId + colIdStart)              = value;
+          colId++;
+        }
+      });
+
+  Kokkos::deep_copy(columnIndices, uniqueColumnIndices);
+  auto exec_space = matrix_execution_space{};
+  Kokkos::sort(uniqueColumnIndices);
+  auto end = Kokkos::Experimental::unique(exec_space, uniqueColumnIndices);
+  auto nUnique =
+      Kokkos::Experimental::distance(Kokkos::Experimental::begin(uniqueColumnIndices), end);
+  auto listColumnIndices = Kokkos::subview(uniqueColumnIndices, Kokkos::pair<int, int>(0, nUnique));
+
+  auto colMap = RCP<Tpetra::Map<LO, GO, NT>>(new Tpetra::Map<LO, GO, NT>(
+      invalid, listColumnIndices, rangeMap->getIndexBase(), rangeMap->getComm()));
```

Comment:
I am not sure I understand what's happening here. Don't you already have this info?

Reply:
I'll opt to use `Tpetra::Details::makeColMap`, since this code section was buggy/error-prone.